### PR TITLE
planner: implement projection and table dual in cascades

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Not a problem, this is part of the new cascades planner. This PR is split from https://github.com/pingcap/tidb/pull/11566.

### What is changed and how it works?

Add implementation rules for `LogicalProjection` and `LogicalTableDual`. After this PR, we can run `select 1` query when cascades planner is enabled.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

N/A

Related changes

N/A

### CODE REVIEW

File: planner/core/plan.go [hunk 1]
```diff
@@ -312,7 +312,8 @@ func (p *basePhysicalPlan) SetChild(i int, child PhysicalPlan) {
 	p.children[i] = child
 }
 
-func (p *basePlan) context() sessionctx.Context {
+// Context implements Plan Context interface.
+func (p *basePlan) Context() sessionctx.Context {
```
Comment (zz-jason) [inline planner/core/plan.go]: `Context` is a little confusion, how about renaming it to `SCtx()` to be consist with other places?

File: planner/core/logical_plans.go [hunk 2]
```diff
@@ -217,18 +217,18 @@ type LogicalProjection struct {
...
-	// calculateNoDelay indicates this Projection is the root Plan and should be
+	// CalculateNoDelay indicates this Projection is the root Plan and should be
 	// calculated without delay and will not return any result to client.
 	// Currently it is "true" only when the current sql query is a "DO" statement.
 	// See "https://dev.mysql.com/doc/refman/5.7/en/do.html" for more detail.
-	calculateNoDelay bool
+	CalculateNoDelay bool
 
-	// avoidColumnRef is a temporary variable which is ONLY used to avoid
+	// AvoidColumnRef is a temporary variable which is ONLY used to avoid
```
Comment (lamxTyler) [inline planner/core/logical_plans.go]: `AvoidColumnEvaluator`?

File: planner/implementation/datasource.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+package implementation
+
+import (
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/planner/memo"
+)
+
+// TableDualImpl implementation of PhysicalTableDual.
+type TableDualImpl struct {
```
Comment (lzmhhh123) [inline planner/implementation/datasource.go]: What about move `ProjectionImpl` out of `base.go` to create a file like this.

**Pull request merged**